### PR TITLE
Decorrelation and parallelization of recursive and materialized CTEs

### PR DESCRIPTION
**Pull request by kryonix**
This PR contains several CTE-related improvements.

---
1. With this PR it is possible to decorrelate recursive, and materialized CTEs. This now enables execution of queries like:

```SQL
SELECT x, y, (WITH RECURSIVE t(z) AS MATERIALIZED (
  SELECT x + y
    UNION ALL
  SELECT z + 1
  FROM   (WITH RECURSIVE g(a) AS MATERIALIZED (
          SELECT t.z
          FROM   t
            UNION ALL
          SELECT g.a + (x + y) / 2
          FROM   g
          WHERE  g.a < 3)
          SELECT * FROM g) AS t(z)
  WHERE z < 5
) SELECT sum(z) FROM t) AS z
FROM   generate_series(1,2) AS _(x), generate_series(1,2) AS __(y);
```

---

2. The execution method of materialized CTEs has been simplified. Instead of mimicking the mechanism of recursive CTEs, they now work similar to the materialization phase of delim joins.  This enables parallelization of materialized CTEs.

---

3. Recursive CTEs are now a parallel sink.

### CODE REVIEW

**Comment by kryonix** [comment]
The `MarkSubtreeCorrelated` logic was too aggressive, which resulted in suboptimal plans.

File: src/include/duckdb/planner/subquery/flatten_dependent_join.hpp [hunk 1]
```diff
@@ -25,6 +25,9 @@ struct FlattenDependentJoins {
 	//! has_correlated_expressions map.
 	bool DetectCorrelatedExpressions(LogicalOperator *op, bool lateral = false, idx_t lateral_depth = 0);
 
+	//! Mark entire subtree of Logical Operators as correlated by adding them to the has_correlated_expressions map.
+	bool MarkSubtreeCorrelated(LogicalOperator *op);
```
Comment (Mytherin) [inline src/include/duckdb/planner/subquery/flatten_dependent_join.hpp:29]: nit: `LogicalOperator &`

File: src/planner/subquery/flatten_dependent_join.cpp [hunk 2]
```diff
@@ -576,10 +603,73 @@ unique_ptr<LogicalOperator> FlattenDependentJoins::PushDownDependentJoinInternal
...
+		RewriteCorrelatedExpressions recursive_rewriter(this->base_binding, correlated_map, lateral_depth + 1, true);
+		recursive_rewriter.VisitOperator(*plan->children[0]);
+		recursive_rewriter.VisitOperator(*plan->children[1]);
+
+#ifdef DEBUG
+		plan->children[0]->ResolveOperatorTypes();
+		plan->children[1]->ResolveOperatorTypes();
+#endif
+		// we have to refer to the recursive CTE index now
+		base_binding.table_index = setop.table_index;
```
Comment (Mytherin) [inline src/planner/subquery/flatten_dependent_join.cpp:630]: The recursive/materialized CTE code here looks almost identical except for these three lines, perhaps they can be merged into the same switch case with these three lines only executed for `LogicalOperatorType::LOGICAL_RECURSIVE_CTE`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Extremely impressive work! This is awesome.


I have been playing around with it for a bit. In particular, the way I think about correlated subqueries is that they allow you to "parameterize" queries. So we can extract any constant from the query, make it correlated, and it should return the same result. For example:

```sql
SELECT * FROM range(100) t(x) WHERE x < 5;
```

Should return the same result as:

```sql
SELECT t2.* FROM (VALUES (5)) t(_corr), LATERAL (SELECT * FROM range(100) t(x) WHERE x < _corr) t2;
```

With that in mind, I have been taking a few of the recursive CTE queries from the other tests and rewriting them in such a manner to stress test this PR. In general it works very well! I have found the following what I believe to be issues:

##### Aggregates & NULL handling

Taken from `test/sql/cte/recursive_cte_complex_pipelines.test`:

###### Original Query
```sql
CREATE TABLE a AS SELECT * FROM range(100) t1(i);
WITH RECURSIVE t AS
(
	SELECT 1 AS x
UNION
	SELECT SUM(x) AS x
	FROM t, a
	WHERE x < 1000000
)
SELECT * FROM t ORDER BY 1 NULLS LAST;
┌─────────┐
│    x    │
│  int32  │
├─────────┤
│       1 │
│     100 │
│   10000 │
│ 1000000 │
│    NULL │
└─────────┘
```

###### Lateral

```sql
SELECT t2.*
FROM (VALUES (1000000)) t(_corr), LATERAL (
WITH RECURSIVE t AS
(
	SELECT 1 AS x
UNION
	SELECT SUM(x) AS x
	FROM t, a
	WHERE x < _corr
)
SELECT * FROM t) t2
ORDER BY 1 NULLS LAST;
┌─────────┐
│    x    │
│  int32  │
├─────────┤
│       1 │
│     100 │
│   10000 │
│ 1000000 │
└─────────┘
```

The result is correct - except it seems that the `NULL` value has disappeared.

##### Correlated Query in Recursive CTE

###### Original Query
```sql
WITH RECURSIVE t AS
(
	SELECT 1 AS x
UNION
	SELECT (SELECT t.x+t2.x FROM t t2 LIMIT 1) AS x
	FROM t
	WHERE x < 10
)
SELECT * FROM t ORDER BY 1 NULLS LAST;
┌───────┐
│   x   │
│ int32 │
├───────┤
│     1 │
│     2 │
│     4 │
│     8 │
│    16 │
└───────┘
```

###### Lateral

```sql
SELECT t2.*
FROM (VALUES (10)) t(_corr), LATERAL (
WITH RECURSIVE t AS
(
	SELECT 1 AS x
UNION
	SELECT (SELECT t.x+t2.x FROM t t2 LIMIT 1) AS x
	FROM t
	WHERE x < _corr
)
SELECT * FROM t) t2
ORDER BY 1 NULLS LAST;
-- ABORT THROWN BY INTERNAL EXCEPTION: Attempted to access index 2 within vector of size 2
```


##### Nested Recursive CTE

Taken from `test/sql/cte/test_nested_recursive_cte.test`

###### Original Query

```sql
WITH RECURSIVE collatz(x, t, steps) AS
(
  SELECT x, x, 0
  FROM   (WITH RECURSIVE n(t) AS (SELECT 1 UNION ALL SELECT t+1 FROM n WHERE t < 10) SELECT * FROM n) AS _(x)
    UNION ALL
  (SELECT x, CASE WHEN t%2 = 1 THEN t * 3 + p ELSE t / 2 END, steps + p
   FROM   collatz, (WITH RECURSIVE n(t) AS (SELECT 1 UNION ALL SELECT t+1 FROM n WHERE t < 1) SELECT * FROM n) AS _(p)
   WHERE  t <> 1)
)
SELECT * FROM collatz WHERE t = 1
ORDER BY x;
```

###### Lateral

```sql
SELECT t2.*
FROM (VALUES (1)) t(_corr), LATERAL (
WITH RECURSIVE collatz(x, t, steps) AS
(
  SELECT x, x, 0
  FROM   (WITH RECURSIVE n(t) AS (SELECT _corr UNION ALL SELECT t+_corr FROM n WHERE t < 10) SELECT * FROM n) AS _(x)
    UNION ALL
  (SELECT x, CASE WHEN t%2 = _corr THEN t * 3 + p ELSE t / 2 END, steps + p
   FROM   collatz, (WITH RECURSIVE n(t) AS (SELECT _corr UNION ALL SELECT t+_corr FROM n WHERE t < _corr) SELECT * FROM n) AS _(p)
   WHERE  t <> _corr)
)
SELECT * FROM collatz WHERE t = _corr
ORDER BY x
) t2;
-- ABORT THROWN BY INTERNAL EXCEPTION: Assertion triggered in file "/Users/myth/Programs/duckdb/src/execution/column_binding_resolver.cpp" on line 151: expr.depth == 0
```

I kind of feel bad even reporting this one :)

**Comment by kryonix** [comment]
Thanks for having a look and for the test cases @Mytherin. I agree, correlation should not change the result. These issues definitively have to be addressed.

File: src/planner/subquery/flatten_dependent_join.cpp (hunk 2 above)
Comment (kryonix) [reply src/planner/subquery/flatten_dependent_join.cpp]: Absolutely right. I fixed that.

**Comment by kryonix** [comment]
All issues are fixed except for the missing `NULL` value in:

```
SELECT t2.*
FROM (VALUES (1000000)) t(_corr), LATERAL (
WITH RECURSIVE t AS
(
	SELECT 1 AS x
UNION
	SELECT SUM(x) AS x
	FROM t, a
	WHERE x < _corr
)
SELECT * FROM t) t2
ORDER BY 1 NULLS LAST;
```

Not quite sure (yet) what causes this.

File: src/include/duckdb/planner/subquery/flatten_dependent_join.hpp (hunk 1 above)
Comment (kryonix) [reply src/include/duckdb/planner/subquery/flatten_dependent_join.hpp]: This is fixed.

**Comment by Mytherin** [comment]
Thanks for the fixes! It looks like there's an issue with the single-file compilation (`DISABLE_UNITY=1 make debug`)

**Pull request merged**